blog/2021-05-29-using-rust.md

```diff
@@ -0,0 +1,74 @@
+---
+title: <undecided>
+author: Rohan
+github: RohanGautam
+description: Explore the use of rust in FoodDX, and go over key architectural decisions
+tags: rust
+---
+
+# Introduction to the product FoodDX
+
+## Product intro
+
+[FoodDX](https://www.fooddx.com/) is a service which helps people get insights on how to improve their diet with the help from our proprietary AI technology and personalized feedback from our nutrition experts. In it's current stage, it helps score images of food taken through an app, and gives it a score from 1-5, and also provides personalized tips for the food.
+
+It's a big project with a lot of components - the AI models, the app, and the backend infrastructure which handles it all. We'll be taking a look at the backend in this article.
+
+## Lifecycle of an image
+
+Before we dive into the infrastructure, it would help to take a look at what an image goes through once it enters our system.
+
+- The image is assigned a UUID.
+- The image is uploaded to a `s3` bucket.
+- The image is then downloaded, preprocessed, run through our models, and a food tip is generated for the image.
+- The food tip, inference results and the image hash form the final response.
+
+## Stack decisions: Rust and haskell
+
+Haskell is used for the client facing API, and it's for other ad-hoc tasks such as reading/writing to a database among others.
+
+Rust is used for image preprocessing, model inference and sending the results back. Sending of results was done differently in the two approaches outlined below. Rust was chosen for it's high efficiency, small executable footprint, and absence of a garbage collector. It also had strong type system, speed & relatively actively maintained Tensorflow (client) library.
+
+# Introduction and pitfalls of the existing architecture
+
+![](/images/v2-arch-diagram.png)
+The internal organization of the rust service in this architecture is outlined above.
+There were 3 main parts, all running concurrently on 3 separate tokio runtimes - namely polling sqs, preprocessing and inferring from the images, and cleanup tasks.
```

Comment:
I would advise a quick intro to Rust async runtimes 
OR 
We assume the reader has to have some basic knowledge of the Rust ecosystem
OR 
We link to `tokio.rs` webpage/documentation that explains briefly what tokio is 
OR 
We can add a quick `Note` type thing with a brief explanation of `tokio` (doesn't have to be written by us)

Reply:
SQS -> Link to AWS SQS Documentation along with a one-liner explaining what SQS is 
Pre-processing -> One-liner explaining that our ML models expect small/fixed image dimensions
Inferring -> I think this is called `running inference on` or something. Can ask Matthew to review that line or ask him for what is the correct terminology. 
Cleanup tasks -> Explain what "cleanup tasks" are there to be done (can be short) and that since the result is not dependent on them, they can be done after we return the result or write the result out to Redis (depending on how much detail we wish to provide).

Reply:
thanks for the input! implemented them in an upcoming commit.

Regarding preprocessing, I think it's understood that image classification models take a fixed length image tensor, it's fairly standard to assume that preprocessing would/does involve that imo. I don't think we need to justify the preprocessing 🤔

Reply:
feel free to mark as unresolved if need be!